zcash_client_backend/src/data_api/mem_wallet.rs

```diff
@@ -379,12 +378,98 @@ impl WalletWrite for MemoryWalletDb {
+        for block in blocks.into_iter() {
+            let mut transactions = HashMap::new();
+            for transaction in block.transactions().into_iter().cloned() {
+                let txid = transaction.txid();
+                let account_id = 0; // TODO: Assuming the account is 0, handle this accordingly.
+                let ufvk = self
+                    .accounts
+                    .get(&account_id)
+                    .ok_or(Error::AccountUnknown(0))?
+                    .ufvk
+                    .sapling()
+                    .ok_or(Error::ViewingKeyNotFound(0))?;
+                let nk = ufvk.to_nk(Scope::External);
+
+                // Insert the Sapling nullifiers of the spent notes into the `sapling_spends` map.
+                transaction.sapling_outputs().iter().map(|o| {
+                    let nullifier = o.note().nf(&nk, o.note_commitment_tree_position().into());
+                    // TODO: Populate the bool field properly.
+                    self.sapling_spends.entry(nullifier).or_insert((txid, true));
+                });
+
+                #[cfg(feature = "orchard")]
+                // Insert the Orchard nullifiers of the spent notes into the `orchard_spends` map.
+                transaction.orchard_outputs().iter().map(|o| {
+                    if let Some(nullifier) = o.nf() {
+                        self.orchard_spends
+                            .entry(*nullifier)
+                            .or_insert((txid, true));
+                    }
+                });
+
+                // Add frontier to the sapling tree
+                self.sapling_tree.insert_frontier(
+                    from_state.final_sapling_tree().clone(),
+                    Retention::Checkpoint {
+                        id: from_state.block_height(),
+                        is_marked: false,
+                    },
+                );
+
+                #[cfg(feature = "orchard")]
+                // Add frontier to the orchard tree
+                self.orchard_tree.insert_frontier(
+                    from_state.final_orchard_tree().clone(),
+                    Retention::Checkpoint {
+                        id: from_state.block_height(),
+                        is_marked: false,
+                    },
+                );
+
+                // TODO: Is `self.tx_idx` field filled with all the transaction ids from the scanned blocks ?
+                self.tx_idx.insert(txid, block.block_height);
+                transactions.insert(txid, transaction);
+            }
+
+            let memory_block = MemoryWalletBlock {
+                height: block.block_height,
+                hash: block.block_hash,
+                block_time: block.block_time,
+                transactions,
+                // TODO: Add memos
+                memos: HashMap::new(),
+            };
+
+            self.blocks.insert(block.block_height, memory_block);
+
+            // Add the sapling commitments to the sapling tree.
```

Comment:
Is this still needed now that we added the frontier?

Reply:
Yes; the frontier just provides a starting point, but the note commitments for each note in the block, and the commitment for the block-end checkpoint, must also be added.

Reply:
Thanks, i added the orchard version of this at https://github.com/zcash/librustzcash/pull/1298/commits/370a4aa7e522faca1b3bc56ae9c7087fa7a6ac91